vi.py

```diff
@@ -0,0 +1,224 @@
+            torch.tensor(a_log_sigma_init)
+        )
+
+        self.in_features = in_features
+        self.hidden_features = hidden_features
+        self.depth = depth
+        self.kl_scaling = kl_scaling
+
+
+
+    def _forward(self, g, x):
+        """ Internal forward.
+
+        Parameters
+        ----------
+        g : dgl.DGLGraph
+            Input graph.
+
+        x : torch.Tensor
+            Input features. (n_nodes, in_features)
+
+        Returns
+        -------
+        dgl.DGLGraph
+            Graph with perturbed weights annotated.
+
+        """
+        g = g.local_var()
+
+        g.apply_edges(
+            lambda edges: {
+                **{
+                    "a0": torch.distributions.Normal(
+                        self.a_mu, self.a_log_sigma,
+                        ).rsample([g.number_of_edges(), self.in_features]),
+                },
+                **{
+                    "a%s" % idx: torch.distribution.Normal(
+                        self.a_mu, self.a_log_sigma,
+                        ).rsample([g.number_of_edges(), self.hidden_features])
+                    for idx in range(1, self.depth)
+                }
+            }
+        )
+
+        for idx in range(self.depth):
+            g.edata["a"] = g.edata["a%s" % idx]
+            x = getattr(self, "gn%s" % idx)(g, x)
+            x = getattr(self, "bn%s" % idx)(x)
+            x = self.activation(x)
+
+        g.ndata["x"] = x
+        return g, x
+
+    @staticmethod
+    def condition(self, x):
+        """ Condition (static method. )
+
+        Takes a set of parameters and return a distribution.
+
+        Parameters
+        ----------
+        x : torch.Tensor (n_graphs, out_features)
+
+        Returns
+        -------
+        torch.distributions.Categorical
+
+        """
+        return torch.distributions.categorical.Categorical(logits=x)
+
+    def loss(self, g, x, y, n_samples=1, mask=None):
+        """ Training loss.
+
+        Parameters
+        ----------
+        g : dgl.DGLGraph
+            Input graph.
+
+        x : torch.Tensor
+            Input features.
+
+        y : torch.Tensor
+            Reference value.
+
+        n_samples : int
+            Number of samples to consider.
+
+        Returns
+        -------
+        torch.Tensor : loss
+
+        """
+        # initialize losses
+        losses = []
+
+        for _ in range(n_samples):
+            # get input graph
+            # x.shape = (n_graphs, out_features)
+            g, x = self._forward(g, x)
+            x = x[mask]
+
+            # posterior distribution
+            p_y_given_x_z = self.condition(x)
+
+            # prior
+            p_a = self.a_prior
+
+            # variational posterior
+            q_a = torch.distributions.Normal(
+                loc=self.a_mu,
+                scale=self.a_log_sigma.exp()
+            )
+
+            # compute elbo
+            elbo = p_y_given_x_z.log_prob(y[mask]).mean()\
```

Comment:
Ok, let's consider the math for a second.

The variables Z, are they global or local? I.e. do we sample them separately for each datapoint, or just for each batch/forward-pass?

The general math if Z is global (and simplifying the multiple rounds  tensor thingie by writing capital Z) is:

logp(Y|X) >= log [ [Prod_i^{N} p(y_i| x_i,Z) ] * p(Z) / q(Z)] ,

Where N is the size of the DATASET, not the batch. So if we have minibatches we need to scale appropriately.
Note that Z is global and not inside the product over N.

IF Z is sampled separately per datapoint, we get:
logp(Y|X) >= log [ [Prod_i^{N} p(y_i| x_i,Z_i) * p(Z_i) / q(Z_i)] .

Note the difference that Z is now indexed by i and inside the product, which means it is a joint sum over the dataset.

What is currently implemented will not work in any way since it is neither of all these cases.

Reply:
- parameters: mu \in R(1); log_sigma \in R(1)
- sampling: sample two numbers for each feature, each edge, and each graph

Reply:
- name of the model: STAG_VI(R)

Reply:
in the loss function in L190, since it's broken down into different message passing rounds $Z$ would have the dimension: (n_edges, n_features)

Reply:
E: number of edges, C: number of features,
- STAG_VI(R^C): parameters mu \in R(C); log_sigma \in R(C); Z \in (E, C), independent in each message passing round;
- STAG_VI(R^E): neural network determines mu \in R(E); log_sigma \in R(E); Z \in (E, C), independent in each mp round;
- STAG_VI(R^(E \times C)): neural network determines mu \in R(E \times C); log_sigma \in R(E \times C); Z in Z \in (E, C)

Reply:
replace `len(losses)` in L198 with `n_samples`

Reply:
don't overwrite `g`